include/aspect/particle/property_pool.h

```diff
@@ -0,0 +1,100 @@
+/*
+ Copyright (C) 2016 by the authors of the ASPECT code.
+
+ This file is part of ASPECT.
+
+ ASPECT is free software; you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation; either version 2, or (at your option)
+ any later version.
+
+ ASPECT is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with ASPECT; see the file doc/COPYING.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef __aspect__particle_property_pool_h
+#define __aspect__particle_property_pool_h
+
+#include <aspect/global.h>
+
+//#include <aspect/particle/particle.h>
+#include <deal.II/base/array_view.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    using namespace dealii;
+
+    /**
+     * This class manages the memory space in which particles store their
+     * properties. Because this is dynamic memory and every particle needs the
+     * same amount it is more efficient to let this be handled by a central
+     * manager that does not need to allocate/deallocate memory every time a
+     * particle is constructed/destroyed.
+     */
+    class PropertyPool {
+      public:
+        struct Handle {
+          public:
+          Handle()
+          :
+            index(numbers::invalid_unsigned_int)
+          {}
+
+          Handle(const std::size_t index)
+          :
+            index(index)
+          {}
+
+          Handle &operator= (const Handle &handle)
+          {
+            index = handle.index;
+            return *this;
+          }
+
+          bool is_valid() const
+          {
+            return index != numbers::invalid_unsigned_int;
+          }
+
+          std::size_t index;
+        };
+
+        PropertyPool (const unsigned int n_properties_per_slot);
+
+        Handle allocate_properties_array ();   // return a handle to one slot
+        void deallocate_properties_array (Handle &);
+
+        ArrayView<double> get_properties (const Handle &);  // translate to actual address
+
+        void
+        consolidate_memory ();
+
+        /**
+         * Reserves the dynamic memory needed for storing the properties of
+         * @p size particles.
+         */
+        void reserve(const unsigned int size);
+
+      private:
+        unsigned int n_properties;
+        std::size_t next_free_index;
+        std::size_t next_free_memory_location;
+
+        std::vector<double> memory_pool;    // size == N_slots * n_properties_per_slot;
+
+        // possibly a translation table from index to memory pool location
+        std::vector<std::size_t> translation_table; // size == N_slots
```

Comment:
Two comments here:
1/ I think the correct inner type is probably `std::vector<double>::size_type` because that's what `memory_pool` uses to index into its array. (Likely `std::size_t`, but let's make it explicit.)

1/ To make this efficient, you will need to keep a list of which slots in `memory_pool` are currently used and which are not. You can in principle gather this information by walking `translation_table` and storing which slots you have seen (assuming that all unused indices in `translation_table` are properly invalidated). But this is not efficient if you want to find an unused slot. My thought was to maybe use an `IndexSet` that represents the indices of all slots currently unused. Allocating a slot would then require finding any element of the index set and then deleting this element from the index set (I'm currently not sure whether we support this last step efficiently), whereas de-allocating a slot would require adding this index back into the index set. (I've also thought about keeping the inverse, i.e., the set of slots that are *used* in such an `IndexSet`, but I think that has no advantage.)